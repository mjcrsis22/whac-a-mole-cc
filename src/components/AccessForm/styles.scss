.access-form {
  display: flex;
  flex-direction: column;

  .form-control {
    margin-bottom: 2rem;
  }
}
