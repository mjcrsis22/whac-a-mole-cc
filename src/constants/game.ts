export enum GAME_LEVEL {
  LOW = 'LOW',
  MEDIUM = 'MEDIUM',
  HARD = 'HARD',
}

export interface GameSetting {
  id: GAME_LEVEL;
  spotRate: number;
  points: number;
}

export const GAME_SETTINGS_BY_LEVEL: {
  [key in keyof typeof GAME_LEVEL]: GameSetting;
} = {
  [GAME_LEVEL.LOW]: {
    id: GAME_LEVEL.LOW,
    spotRate: 1000,
    points: 10,
  },
  [GAME_LEVEL.MEDIUM]: {
    id: GAME_LEVEL.MEDIUM,
    spotRate: 750,
    points: 20,
  },
  [GAME_LEVEL.HARD]: {
    id: GAME_LEVEL.HARD,
    spotRate: 500,
    points: 30,
  },
};
