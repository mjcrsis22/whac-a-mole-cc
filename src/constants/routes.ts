export enum APP_ROUTES {
  HOME = '/home',
  GAME = '/game',
}
