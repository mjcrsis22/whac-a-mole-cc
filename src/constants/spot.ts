export enum SPOT_STATES {
  inactive,
  active,
}

export const SPOT_POSITIONS = [
  { left: '5%', top: '0%' },
  { left: '5%', top: '72%' },
  { left: '30%', top: '15%' },
  { left: '30%', top: '60%' },
  { left: '45%', top: '40%' },
  { left: '60%', top: '15%' },
  { left: '60%', top: '60%' },
  { left: '85%', top: '0%' },
  { left: '85%', top: '72%' },
];
