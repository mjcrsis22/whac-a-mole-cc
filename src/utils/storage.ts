export enum STORE_KEYS {
  USERNAME = 'wam_username',
}

export const storeData = (key: string, value: string): void => {
  localStorage.setItem(key, value);
};

export const getStoredData = (key: string): string | null => {
  return localStorage.getItem(key);
};

export const removeStoredData = (key: string): void => {
  localStorage.removeItem(key);
};
