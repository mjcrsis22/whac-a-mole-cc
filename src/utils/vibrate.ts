export const vibrate = (pattern: VibratePattern): boolean => {
  return window.navigator?.vibrate?.(pattern);
};
