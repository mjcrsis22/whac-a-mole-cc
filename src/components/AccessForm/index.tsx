import { useState, FormEvent } from 'react';
import { useNavigate } from 'react-router-dom';
import { storeData, STORE_KEYS } from '../../utils/storage';
import { APP_ROUTES } from '../../constants/routes';
import './styles.scss';

function AccessForm() {
  const navigate = useNavigate();
  const [username, setUsername] = useState('');

  const handleUsernameChange = (event: FormEvent<HTMLInputElement>) => {
    setUsername(event.currentTarget.value);
  };

  const handleSubmit = (event: FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    event.stopPropagation();
    storeData(STORE_KEYS.USERNAME, username);
    navigate(APP_ROUTES.GAME);
  };

  return (
    <form
      id="access-form"
      className="access-form"
      noValidate
      onSubmit={handleSubmit}
    >
      <input
        type="text"
        name="username"
        id="access-username"
        className="form-control"
        maxLength={15}
        placeholder="Name"
        onChange={handleUsernameChange}
      />
      <input
        type="submit"
        id="access-submit-button"
        className="button"
        value="Join"
        disabled={!username}
      />
    </form>
  );
}

export default AccessForm;
