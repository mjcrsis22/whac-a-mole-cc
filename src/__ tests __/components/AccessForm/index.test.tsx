import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { STORE_KEYS, storeData } from '../../../utils/storage';
import { APP_ROUTES } from '../../../constants/routes';
import AccessForm from '../../../components/AccessForm';

const mockedNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: jest.fn(() => mockedNavigate),
}));

jest.mock('../../../utils/storage', () => ({
  ...jest.requireActual('../../../utils/storage'),
  storeData: jest.fn(),
}));

describe('AccessForm component', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  test('Renders correctly', () => {
    const { container } = render(<AccessForm />);
    expect(container).toMatchSnapshot();
  });

  test('Enables submit button when username is written', async () => {
    const result = render(<AccessForm />);

    const button = result.getByRole('button');
    expect(button).toBeDisabled();

    const input = result.getByRole('textbox');
    await userEvent.type(input, 'MiNombre');

    expect(button).not.toBeDisabled();
  });

  test('Calls storeData and navigate on submit', async () => {
    const result = render(<AccessForm />);

    const input = result.getByRole('textbox');
    await userEvent.type(input, 'MiNombre');

    const button = result.getByRole('button');
    await userEvent.click(button);

    expect(storeData).toHaveBeenCalledWith(STORE_KEYS.USERNAME, 'MiNombre');
    expect(mockedNavigate).toHaveBeenCalledWith(APP_ROUTES.GAME);
  });
});
