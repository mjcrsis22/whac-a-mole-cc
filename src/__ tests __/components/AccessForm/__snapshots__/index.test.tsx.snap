// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccessForm component Renders correctly 1`] = `
<div>
  <form
    class="access-form"
    id="access-form"
    novalidate=""
  >
    <input
      class="form-control"
      id="access-username"
      maxlength="15"
      name="username"
      placeholder="Name"
      type="text"
    />
    <input
      class="button"
      disabled=""
      id="access-submit-button"
      type="submit"
      value="Join"
    />
  </form>
</div>
`;
